File: leetcode/487. Max Consecutive Ones II.py
class Solution:
    def findMaxConsecutiveOnes(self, nums: List[int]) -> int:
        dp = [[0, 0] for _ in range(len(nums))]
        dp[0] = [nums[0], 1]
        
        res = 1
        
        for i in range(1, len(nums)):
            if nums[i] == 0:
                dp[i][0] = 0
                dp[i][1] = dp[i-1][0] + 1
            else:
                dp[i][0] = dp[i-1][0] + 1
                dp[i][1] = dp[i-1][1] + 1
            res = max(res, *dp[i])
            
        return res

File: leetcode/487. Max Consecutive Ones II.cpp
class Solution {
public:
    int findMaxConsecutiveOnes(vector<int>& nums) {
        int s00 = nums[0]; // not do
        int s01 = 1; // do
        
        int s10 = 0; // not do
        int s11 = 0; // do
        
        int res = 1;
        
        for (size_t i = 1; i < nums.size(); ++i) {
            if (nums[i] == 0) {
                s10 = 0;
                s11 = s00 + 1;
            } else {
                s10 = s00 + 1;
                s11 = s01 + 1;
            }
            s00 = s10;
            s01 = s11;
            
            res = max({res, s00, s01});
        }
        
        return res;
        
    }
};
